Rspec unitテスト(コントローラ)

pokotyan opened the pull request:
# WHAT
- コントローラのテスト
- テスト中に見つけたバグの修正
# WHY

堅牢で信頼性の高いアプリケーションを作るため

pokotyan committed:
create pull request

pokotyan committed:
add test GET#index for prototypes_controller

pokotyan committed:
bug fix for user edit

以前はupdate_paramsにそもそもpasswordを含めてなかった。で、passwordを空白（””）の状態でsubmitしてもちゃんとユーザー情報は更新されてた。
ブランチ「RspecUnitテスト（モデル）」にて
  validates :password, length: { minimum: 7 }
というバリデーションを追加した。
そのせいで、ユーザー情報を更新する際はpasswordを必ず入力しないとバリデーションに引っかかって更新されない状態になった。
なので、update_paramsに:passwordを追加した。
また、deviseの仕様のパスワード更新時の自動的ログアウトを回避するため、
　sign_in(current_user, bypass: true)
を追加。
また、アバター写真表示のためのクラス（.js-image-preview）が当たってなかったので追加。

pokotyan committed:
質問するために現在の状態をpushするためのコミット

pokotyan commented:
POST #createのテストが通りません。deviseのログインモジュールが悪いのか、paramsが悪いのかどちらかのような気がしますが、アドバイスをお願いいたします。
- テスト結果
  Failures:
  
  1) PrototypesController with user login POST #create with valid attributes saves the new prototype in the database
     Failure/Error: @prototype = current_user.prototypes.build(create_params)
  
     NoMethodError:
       undefined method `[]' for #<Tempfile:0x007fb14de9e980>
     # ./app/controllers/prototypes_controller.rb:21:in`create'
     # ./spec/controllers/prototypes_controller_spec.rb:51:in `block (6 levels) in <top (required)>'
     # ./spec/controllers/prototypes_controller_spec.rb:50:in`block (5 levels) in <top (required)>'
     # ./spec/support/database_cleaner.rb:9:in `block (3 levels) in <top (required)>'
     # ./spec/support/database_cleaner.rb:8:in`block (2 levels) in <top (required)>'
- params
  
  > attributes_for(:prototype_with_main_image, images_attributes: attributes_for(:main_image))
  > => {:title=>"test",
  > :catch_copy=>"test_catch_copy",
  > :concept=>"test_concept",
  > :images_attributes=>
  > {:image=>
  >   #<Rack::Test::UploadedFile:0x007febcf04a9f8
  >    @content_type="image/jpg",
  >    @original_filename="sample.jpg",
  >    @tempfile=
  >     #<File:/var/folders/_q/jzlx81kd0kxf7xc24cbh5_b80000gn/T/sample.jpg20161018-19148-ol5rpy>>,
  > :status=>0}}

KoichiOzaki commented:
エラーを直接解釈すると、carrier wave側で画像ファイルを受けとれてないエラーですが、、
下記２点を確認することで問題の切り分けができそうです。
・過去にdeviseを介したログイン機能をテストで使用し、うまく通ったことはあるか？
・prototypeのテスト用イメージを正しく、specファイルから読み込めたことはあるか？
特にログインがこのあたりの処理には必要なはずですが、ログイン処理がspecファイルに記述されているか気になります。
deviseを使用したログイン機能をspecから使用する場合は、通常はdeviseのTestHelpersを読み込む設定が必要です。

KoichiOzaki commented:
私も詳細を動かしながら確認しないことには原因を断言できないのですが、
まずは
・imageファイルを読み込めていないことが原因なのか
・ログインができていないことが原因なのか
この２つを切り分ける必要があるので、調査してみてください。

pokotyan committed:
fix bug

ログインしていない状態でprototypeのshowページにアクセスした際、current_user.likesがnilでエラーになったので、ログインしていない状態でlikeボタンをクリックしたらログインページに遷移するように修正。

pokotyan committed:
add test POST#create for prototype_controller

pokotyan committed:
add test GET#edit for prototype_controller

pokotyan committed:
add test PATCH#update for prototype_controller

pokotyan committed:
add test DELETE#destroy for prototype_controller

pokotyan committed:
add test with_out_user_login for prototype_controller

pokotyan committed:
add test GET #show for users_controller

pokotyan committed:
add test GET #edit for users_controller

pokotyan committed:
add test PATCH #update for users_controller

pokotyan committed:
add test without_user_login for users_controller

pokotyan commented:
![2016-10-21 19 50 28](https://cloud.githubusercontent.com/assets/20484619/19596071/c47a6e38-97c7-11e6-9cc7-4b7199b03316.png)
![2016-10-21 19 52 03](https://cloud.githubusercontent.com/assets/20484619/19596096/e9a74cda-97c7-11e6-84de-d5c9151b2743.png)

hand12 commented on spec/controllers/prototypes_controller_spec.rb:
```diff
@@ -0,0 +1,314 @@
+      #deviseのログイン(login_user)を事前にしておかないと、コントローラのcurrent_user.prototypes〜の部分で
+      #undefined method `prototypes' for nil:NilClass になる。before_action :authenticate_user!を設定してなくてもnilになる。
+      login_user
+      context "with valid attributes" do
```
login_userはログインしている状態を作り出すためにやっているので、contextで"ログインしている状態のとき"というのを作り出し、最初にlogin_userを書いてあげる方が無駄なく分かりやすいコードとなります。
（テストなので、繰り返し同じことを書くのは悪くはないのですが、login_userのような記述は最初に書いてあげた方が読みやすいです。)

pokotyan replied:
login_userを上の方に書いて、繰り返し記載していたlogin_userを一つのみに修正しました。

hand12 commented on spec/controllers/prototypes_controller_spec.rb:
```diff
@@ -0,0 +1,314 @@
+          prototype: attributes_for(:prototype_with_main_image,
+          user_id: @user.id,
+          images_attributes: { "0": attributes_for(:main_image) }
+          ) }
```
ここをFactoryGirlで生成しなかった理由はなんですか？？
特別な状態のインスタンスを作りたいのであれば、FactoryGirlのtraitというメソッドを使うと、より完結に書くことができます。

hand12 commented on spec/controllers/prototypes_controller_spec.rb:
```diff
@@ -0,0 +1,314 @@
+            title: nil,
+            user_id: @user.id,
+            images_attributes: { "0": attributes_for(:main_image) }
+            ) }
```
ここも上記と同じで、特別な状態のインスタンスを作る場合はFactoryGirlのtraitを使った方がすっきりかけます。

hand12 commented on app/controllers/prototypes_controller.rb:
```diff
@@ -33,11 +34,11 @@ def edit
+      redirect_to edit_prototype_path(@prototype), alert: "プロトタイプの更新に失敗しました"
     end
   end
 
```
今回のプルリクとは関係のないことなので、このブランチでは関係ない作業はしないようにしましょう！
今回は大丈夫ですので次回から気をつけてください！

pokotyan replied:
renderだと変数の内容が引き継がれた状態でeditページに戻ってしまうので、
無効な属性を引き継いだ状態でeditページに戻ることを防ぐため、redirect_toに変えました。

以下のテストを実施する際に気づいたバグの修正でした。
    PATCH #update
      with invalid attributes
        renders the :edit template

hand12 reviewed (commented):
コメントしました！
いくつかコメントしましたのでご確認ください！

pokotyan committed:
refactaling of test

yogoken commented on app/controllers/prototypes_controller.rb:
```diff
@@ -33,11 +34,11 @@ def edit
   end
 
   def update
+    #binding.pry
```
binding.pryは削除しておきましょう！

pokotyan replied:
消しました！

yogoken commented on app/views/users/edit.html.haml:
```diff
@@ -6,13 +6,13 @@
           .user-name.col-md-10
             = f.text_field :name, placeholder: "Username"
-          .user-image.col-md-2
+          .user-image.col-md-2.js-image-preview
```
jsはこちらで作業されましたでしょうか？
次回のみRspecの作業以外の変更は他のブランチ切った後にやっていただいた方がわかりやすいのではと思われます。今回の場合は全く問題ないですよ！

yogoken reviewed (commented):
こちらコメントしました。
ほとんど変える必要はないですね!!
マージしましょう!!

pokotyan committed:
remove line for debug
